Adds replacement CJ clothing models

W3lac3 opened the pull request:
This PR will allow you to replace clothing models in CJ using the `engineLoadDFF` and `engineReplaceModel` functions. The clothing models should be made in the same way as in the original game, with 3 clumps, ripped, normal and fat,

List of model ids:
```lua
afro.dff -> 30000
afrobeard.dff -> 30001
afrotash.dff -> 30002
balaclava.dff -> 30003
bandana.dff -> 30004
bandknots.dff -> 30005
bandmask.dff -> 30006
barefeet.dff -> 30007
baseball.dff -> 30008
bask1.dff -> 30009
baskball.dff -> 30010
bbjack.dff -> 30011
beret.dff -> 30012
biker.dff -> 30013
bikerhelmet.dff -> 30014
boater.dff -> 30015
bowler.dff -> 30016
boxingcap.dff -> 30017
boxingshoe.dff -> 30018
boxingshort.dff -> 30019
cap.dff -> 30020
capandband.dff -> 30021
capback.dff -> 30022
capknit.dff -> 30023
capover.dff -> 30024
capovereye.dff -> 30025
cappolice.dff -> 30026
caprimup.dff -> 30027
capside.dff -> 30028
captruck.dff -> 30029
chinos.dff -> 30030
chinosb.dff -> 30031
chonger.dff -> 30032
chucks.dff -> 30033
coach.dff -> 30034
conv.dff -> 30035
cornrows.dff -> 30036
country.dff -> 30037
countrytr.dff -> 30038
cowboy.dff -> 30039
cs_afro.dff -> 30040
cs_afrobeard.dff -> 30041
cs_afrotash.dff -> 30042
cs_bandana.dff -> 30043
cs_bandknots.dff -> 30044
cs_bandmask.dff -> 30045
cs_bikerhelmet.dff -> 30046
cs_cornrows.dff -> 30047
cs_elvishair.dff -> 30048
cs_flatears.dff -> 30049
cs_flattop.dff -> 30050
cs_gimpmask.dff -> 30051
cs_groovecut.dff -> 30052
cs_hands.dff -> 30053
cs_head.dff -> 30054
cs_highafro.dff -> 30055
cs_jheri.dff -> 30056
cs_mohawk.dff -> 30057
cs_slope.dff -> 30058
cs_tramline.dff -> 30059
cs_wedge.dff -> 30060
cs_zorro.dff -> 30061
denim.dff -> 30062
elvishair.dff -> 30063
eyepatch.dff -> 30064
feet.dff -> 30065
field.dff -> 30066
flatears.dff -> 30067
flattop.dff -> 30068
flipflop.dff -> 30069
garagetop.dff -> 30070
garagetr.dff -> 30071
gimpfeet.dff -> 30072
gimphands.dff -> 30073
gimpleg.dff -> 30074
gimpmask.dff -> 30075
gimptorso.dff -> 30076
glasses01.dff -> 30077
glasses03.dff -> 30078
glasses04.dff -> 30079
glassesx.dff -> 30080
groovecut.dff -> 30081
grouchos.dff -> 30082
hands.dff -> 30083
hatmanc.dff -> 30084
hattiger.dff -> 30085
hawaii.dff -> 30086
head.dff -> 30087
helmet.dff -> 30088
highafro.dff -> 30089
hitop.dff -> 30090
hockeymask.dff -> 30091
hoodjack.dff -> 30092
hoodya.dff -> 30093
hoodyb.dff -> 30094
jeans.dff -> 30095
jheri.dff -> 30096
leather.dff -> 30097
leathertr.dff -> 30098
legs.dff -> 30099
legspants.dff -> 30100
medictr.dff -> 30101
mohawk.dff -> 30102
moto.dff -> 30103
neck.dff -> 30104
neck2.dff -> 30105
neck2_bb.dff -> 30106
neck2_field.dff -> 30107
neck2_hoodya.dff -> 30108
neck2_pimp.dff -> 30109
neck2_shirt.dff -> 30110
neck_bb.dff -> 30111
neck_field.dff -> 30112
neck_hoodya.dff -> 30113
neck_pimp.dff -> 30114
neck_shirt.dff -> 30115
neck_tracky.dff -> 30116
painter.dff -> 30117
pimp.dff -> 30118
pimpshoe.dff -> 30119
pimptr.dff -> 30120
policeshirt.dff -> 30121
policetr.dff -> 30122
shirta.dff -> 30123
shirtb.dff -> 30124
shoe.dff -> 30125
shorts.dff -> 30126
skullycap.dff -> 30127
sleevt.dff -> 30128
slope.dff -> 30129
sneaker.dff -> 30130
suit1.dff -> 30131
suit1tr.dff -> 30132
suit2.dff -> 30133
sweat.dff -> 30134
timber.dff -> 30135
torso.dff -> 30136
tracktr.dff -> 30137
trackytop1.dff -> 30138
tramline.dff -> 30139
trilby.dff -> 30140
tshirt.dff -> 30141
tshirt2.dff -> 30142
valet.dff -> 30143
vest.dff -> 30144
watch.dff -> 30145
watch_hoody.dff -> 30146
wcoat.dff -> 30147
wedge.dff -> 30148
worktr.dff -> 30149
worktrboot.dff -> 30150
zorromask.dff -> 30151
```

W3lac3 committed:
Adds replacement CJ clothing models

TheNormalnij commented on Client/mods/deathmatch/logic/CClientDFF.cpp:
```diff
@@ -101,6 +101,21 @@ void CClientDFF::UnloadDFF()
 
 bool CClientDFF::ReplaceModel(unsigned short usModel, bool bAlphaTransparency)
 {
+    if (usModel >= CLOTHES_MODEL_ID_FIRST && usModel <= CLOTHES_MODEL_ID_LAST)
```
Add `CPlayerClothes::IsValidClothingModelId` for this check

TheNormalnij commented on Client/mods/deathmatch/logic/CClientDFF.cpp:
```diff
@@ -101,6 +101,21 @@ void CClientDFF::UnloadDFF()
+        g_pGame->GetRenderWare()->ClothesAddReplacement(m_RawDataBuffer.data(), usModel - CLOTHES_MODEL_ID_FIRST);
+        return true;
+    }
+
```
Extract this method as `ReplaceClothes(uint32_t model)`
Call it in `DoReplaceModel` 
```c++
        if (CPlayerClothes::IsValidClothingModelId(usModel))
        {
            return ReplaceClothes(usModel);
        }
```

TheNormalnij commented on Client/mods/deathmatch/logic/CClientDFF.cpp:
```diff
@@ -231,6 +246,7 @@ void CClientDFF::RestoreModels()
 
     // Clear the list
     m_Replaced.clear();
+    g_pGame->GetRenderWare()->ClothesRemoveReplacement(m_RawDataBuffer.data());
```
Move it into `InternalRestoreModel` as
```c++
        else if (CPlayerClothes::IsValidClothingModelId(usModel))
        {
            g_pGame->GetRenderWare()->ClothesRemoveReplacement(m_RawDataBuffer.data());
            return;
        }
 ```

TheNormalnij commented:
It looks like it doesn't work unless your model is linked with ingame texture. engineImportTXD doesn't help in this situation.
This feature doesn't work with most mods. 
Can you provide a test resource, please?
Crash resource: [clothes_test_crash.zip](https://github.com/user-attachments/files/18470616/clothes_test_crash.zip)

W3lac3 commented:
> It looks like it doesn't work unless your model is linked with ingame texture. engineImportTXD doesn't help in this situation. This feature doesn't work with most mods. Can you provide a test resource, please? Crash resource: [clothes_test_crash.zip](https://github.com/user-attachments/files/18470616/clothes_test_crash.zip)

From the tests I've done, I've noticed that if the clothe is bigger than the original clothe in the game, it crashes. I was testing with a model of mine, this model is light and wasn't giving me this problem, so I couldn't notice it. At the moment I still have no idea how to solve it, so I've left the PR as a draft until I find a way to solve it.

W3lac3 committed:
Update due to code revision

W3lac3 committed:
Fixes clothing size crash

W3lac3 committed:
Merge remote-tracking branch 'upstream/master' into feature/replace-cj-clothe-model

W3lac3 commented:
The crash when replacing CJ's clothing model has been resolved. Here's an example of a working script: 
[replace_clothes.zip](https://github.com/user-attachments/files/18546490/replace_clothes.zip)

TheNormalnij commented on Client/game_sa/CRenderWareSA.ClothesReplacing.cpp:
```diff
@@ -45,44 +47,120 @@ namespace
+            return nullptr;
+
+        return directory;
+    }
```
You can define CLOTHES_DIRECTORY and avoid this method
```c++
CDirectorySA*       CLOTHES_DIRECTORY = *reinterpret_cast<CDirectorySA**>(0x5A419B);
```

W3lac3 replied:
> You can define CLOTHES_DIRECTORY and avoid this method
> 
> ```c++
> CDirectorySA*       CLOTHES_DIRECTORY = *reinterpret_cast<CDirectorySA**>(0x5A419B);
> ```

I did it through a function, because this address only exists when the game loads the directory.

TheNormalnij replied:
Are you sure?

W3lac3 replied:
> Are you sure?

I made this update in the commits I sent today.

TheNormalnij commented on Client/game_sa/CRenderWareSA.ClothesReplacing.cpp:
```diff
@@ -45,44 +47,120 @@ namespace
+            return nullptr;
+
+        DirectoryInfo* entry = ((DirectoryInfo * (__thiscall*)(CDirectorySA*, const char*))0x532450)(directory, pImgItem->szName);
+
```
You can define `CDirectorySA` as a class and extract this lime as a method.
```c++
DirectoryInfo* entry = CLOTHES_DIRECTORY->GetEntry(pImgItem->szName);
```

TheNormalnij commented on Client/game_sa/CRenderWareSA.ClothesReplacing.cpp:
```diff
@@ -45,44 +47,120 @@ namespace
+
+    DirectoryInfo* GetClothesEntry(ushort usFileId)
+    {
+        SPlayerImgItemArray* pItemArray = (SPlayerImgItemArray*)0x00BC12C0;
```
It's better to define this objects outside of this function with better name

Pirulax commented on Client/game_sa/CRenderWareSA.ClothesReplacing.cpp:
```diff
@@ -45,44 +47,120 @@ namespace
+        uint32         m_nCapacity{};
+        uint32         m_nNumEntries{};
+        bool           m_bOwnsEntries{};
+    };
```
I'm pretty sure these structs already exist somewhere in the code, you should re-use that, instead of defining this here (even if not, you should move it out of this file, it doesn't belong here)

W3lac3 replied:
It doesn't exist, I looked it up before I did it. But then I'll follow what TheNormalnij said, I'll create a CDirectorySA class and leave this structure there. I just haven't done it yet because I'm busy.

TheNormalnij replied:
Indeed. DirectoryInfo repeats tImgFileInfo. You need to move DirectoryInfo to sdk and replace tImgFileInfo

W3lac3 committed:
Create CDirectorySA Class

W3lac3 committed:
Replaces CDirectorySA structure for class in ClothesReplacing

W3lac3 committed:
Merge branch 'master' into feature/replace-cj-clothe-model

TheNormalnij commented on Client/game_sa/CDirectorySA.h:
```diff
@@ -0,0 +1,23 @@
+#include <SharedUtil.IntTypes.h>
```
```suggestion
/*****************************************************************************
 *
 *  PROJECT:     Multi Theft Auto
 *  LICENSE:     See LICENSE in the top level directory
 *  FILE:        game_sa/CDirectorySA.h
 *
 *****************************************************************************/

#pragma once
```

TheNormalnij commented on Client/game_sa/CDirectorySA.cpp:
```diff
@@ -0,0 +1,40 @@
+#include "StdInc.h"
```
```suggestion
/*****************************************************************************
 *
 *  PROJECT:     Multi Theft Auto
 *  LICENSE:     See LICENSE in the top level directory
 *  FILE:        game_sa/CDirectorySA.cpp
 *
 *****************************************************************************/

#include "StdInc.h"
```

W3lac3 committed:
Adds comments to the CDirectorySA.cpp and CDirectorySA.h files

TheNormalnij commented on Client/game_sa/CDirectorySA.cpp:
```diff
@@ -0,0 +1,48 @@
+#include "StdInc.h"
+#include "CDirectorySA.h"
+
+DirectoryInfo* CDirectorySA::GetModelEntry(ushort modelId)
```
Replace all `ushort` and `unsigned short` with uint16_t or std::uint16_t

TheNormalnij commented on Client/game_sa/CDirectorySA.h:
```diff
@@ -0,0 +1,31 @@
+
+#pragma once
+
+struct DirectoryInfo
```
Engine classes should have prefix "SAInterface"

W3lac3 replied:
> Engine classes should have prefix "SAInterface"

The class name should then be: CDirectorySAInterface ? Or do you mean putting SA in the struct ?

TheNormalnij replied:
Yes, CDirectorySAInterface

TheNormalnij commented on Client/game_sa/CRenderWareSA.ClothesReplacing.cpp:
```diff
@@ -28,8 +29,10 @@ namespace
-    std::map<ushort, char*> ms_ReplacementClothesFileDataMap;
-    bool                    bClothesReplacementChanged = false;
+    std::map<ushort, char*>  ms_ReplacementClothesFileDataMap;
+    std::map<ushort, uint16> ms_OriginalStreamingSizesMap;
```
unordered_map

TheNormalnij commented on Client/game_sa/CRenderWareSA.ClothesReplacing.cpp:
```diff
@@ -45,44 +48,67 @@ namespace
+    char*               pReturnBuffer;
+
+
+    uint32_t GetSizeInBlocks(uint32_t size)
```
use size_t

TheNormalnij commented on Client/game_sa/CRenderWareSA.ClothesReplacing.cpp:
```diff
@@ -45,44 +48,67 @@ namespace
+        return (div.quot + (div.rem ? 1 : 0));
+    }
+
+    uint32_t GetSizeInBlocks(uint64_t size)
```
size_t here too

W3lac3 committed:
Merge remote-tracking branch 'upstream/master' into feature/replace-cj-clothe-model

W3lac3 committed:
Update due to code revision

W3lac3 committed:
Merge branch 'master' into feature/replace-cj-clothe-model

botder commented on Client/game_sa/CDirectorySA.cpp:
```diff
@@ -0,0 +1,48 @@
+ *  LICENSE:     See LICENSE in the top level directory
+ *  FILE:        game_sa/CDirectorySA.cpp
+ *
+ *****************************************************************************/
```
```suggestion
/*****************************************************************************
 *
 *  PROJECT:     Multi Theft Auto
 *  LICENSE:     See LICENSE in the top level directory
 *  FILE:        game_sa/CDirectorySA.cpp
 *
 *  Multi Theft Auto is available from https://www.multitheftauto.com/
 *
 *****************************************************************************/
```

botder commented on Client/game_sa/CDirectorySA.h:
```diff
@@ -0,0 +1,31 @@
+ *  LICENSE:     See LICENSE in the top level directory
+ *  FILE:        game_sa/CDirectorySA.h
+ *
+ *****************************************************************************/
```
```suggestion
/*****************************************************************************
 *
 *  PROJECT:     Multi Theft Auto
 *  LICENSE:     See LICENSE in the top level directory
 *  FILE:        game_sa/CDirectorySA.h
 *
 *  Multi Theft Auto is available from https://www.multitheftauto.com/
 *
 *****************************************************************************/
```

botder commented on Client/game_sa/CDirectorySA.h:
```diff
@@ -0,0 +1,31 @@
+    uint16 m_nStreamingSize;
+    uint16 m_nSizeInArchive;
+    char   m_szName[24];
+};
```
```suggestion
struct DirectoryInfoSA
{
    std::uint32_t m_nOffset;
    std::uint16_t m_nStreamingSize;
    std::uint16_t m_nSizeInArchive;
    char          m_szName[24];
};
```

botder commented on Client/game_sa/CDirectorySA.h:
```diff
@@ -0,0 +1,31 @@
+        DirectoryInfoSA* m_pEntries{};
+        uint32         m_nCapacity{};
+        uint32         m_nNumEntries{};
+        bool           m_bOwnsEntries{};
```
```suggestion
        DirectoryInfoSA* m_pEntries{};
        std::uint32_t    m_nCapacity{};
        std::uint32_t    m_nNumEntries{};
        bool             m_bOwnsEntries{};
```

botder commented on Client/game_sa/CRenderWareSA.ClothesReplacing.cpp:
```diff
@@ -45,45 +48,62 @@ namespace
-    int   iReturnFileId;
-    char* pReturnBuffer;
+    DWORD                        FUNC_CStreamingConvertBufferToObject = 0x40C6B0;
+    CDirectorySAInterface*       CLOTHES_DIRECTORY = *reinterpret_cast<CDirectorySAInterface**>(0x5A419B);
```
This offset is inside a function: `CClothesBuilder::LoadCdDirectory`.
Shouldn't this be `0xBC12C0`?

Also: rename the variable to `g_clothesDirectory`

W3lac3 replied:
> This offset is inside a function: `CClothesBuilder::LoadCdDirectory`. Shouldn't this be `0xBC12C0`?
> 
> Also: rename the variable to `g_clothesDirectory`

This address points to the CJ clothing directory. Through it I can even add new clothes without replacing them (I intend to bring this PR later).

botder replied:
It makes no sense to take the pointer from there.
![image](https://github.com/user-attachments/assets/4ac39201-31c1-4211-a9a0-84e4f169d435)

W3lac3 replied:
> It makes no sense to take the pointer from there.

Yes, this variable is loaded into `CClothesBuilder::LoadCdDirectory`. But, the address that points to CJ's clothes directory is the address I used, you can even see in the modloader repository that they use this same address to load CJ's clothes.

botder replied:
You took the address from within an actual game function that uses the true pointer to the clothes directory. Just take the actual address and not an indirection.

W3lac3 replied:
> You took the address from within an actual game function that uses the true pointer to the clothes directory. Just take the actual address and not an indirection.

From the address you mentioned `0xBC12C0` it is not possible to add clothing models or change entry values such as `m_nStreamingSize`. The address I use allows me to make these changes. As you said I shouldn't use this address, I may have to change the PR to draft. I don't really understand reverse engineering, as I've just started studying it and it will take me a while to find the right address for this purpose.

TheNormalnij replied:
Does it work?
 ```c++
 CDirectorySAInterface* g_clothesDirectory = reinterpret_cast<CDirectorySAInterface*>(0xBC12C0)
 ```

W3lac3 replied:
> Does it work?
> 
> ```c++
> CDirectorySAInterface* g_clothesDirectory = reinterpret_cast<CDirectorySAInterface*>(0xBC12C0)
> ```

I'll try it out.

W3lac3 replied:
> Does it work?

I tested it again, it worked. I had just changed the address to this one, so it wasn't working. I was an idiot 😂

botder replied:
> Does it work?
> 
> ```c++
> CDirectorySAInterface* g_clothesDirectory = reinterpret_cast<CDirectorySAInterface*>(0xBC12C0)
> ```

Yes, obviously.

botder commented on Client/game_sa/CRenderWareSA.ClothesReplacing.cpp:
```diff
@@ -45,45 +48,62 @@ namespace
+
+    size_t GetSizeInBlocks(size_t size)
+    {
+        auto div = std::div(size, 2048);
```
Please rename the variable `div` to avoid confusion with `std::div`

W3lac3 committed:
Merge remote-tracking branch 'upstream/master' into feature/replace-cj-clothe-model

W3lac3 committed:
Update due to code revision

botder commented:
I've noticed that the game crashes if CJ is using the clothes that a script is trying to replace. It also crashes if you stop the resource that replaced the clothes.

W3lac3 commented:
> I've noticed that the game crashes if CJ is using the clothes that a script is trying to replace. It also crashes if you stop the resource that replaced the clothes.

Which model are you using? Are you using the script I provided?

W3lac3 committed:
Update clothes directory memory address

botder commented:
I've replaced policeshirt and policetr

W3lac3 commented:
> I've replaced policeshirt and policetr

The models must follow the CJ pattern, with Ripped, Fat and Normal clumps. And the texture should also follow the game's pattern. About the crash when you turn off the resource, I think I know what it is, I'll check now.

W3lac3 committed:
Fix crash when turning off resource

botder commented on Client/game_sa/CRenderWareSA.ClothesReplacing.cpp:
```diff
@@ -28,8 +29,10 @@ namespace
-    std::map<ushort, char*> ms_ReplacementClothesFileDataMap;
-    bool                    bClothesReplacementChanged = false;
+    std::unordered_map<ushort, char*>  ms_ReplacementClothesFileDataMap;
+    std::unordered_map<ushort, uint16> ms_OriginalStreamingSizesMap;
```
```suggestion
    std::unordered_map<ushort, char*>         ms_ReplacementClothesFileDataMap;
    std::unordered_map<ushort, std::uint16_t> ms_OriginalStreamingSizesMap;
```

botder commented on Client/game_sa/CRenderWareSA.ClothesReplacing.cpp:
```diff
@@ -45,45 +48,65 @@ namespace
+    DWORD                        FUNC_CStreamingConvertBufferToObject = 0x40C6B0;
+    CDirectorySAInterface*       g_clothesDirectory = reinterpret_cast<CDirectorySAInterface*>(0xBC12C0);
+    int                          iReturnFileId;
+    char*                        pReturnBuffer;
```
```suggestion
    DWORD FUNC_CStreamingConvertBufferToObject = 0x40C6B0;
    auto  g_clothesDirectory = reinterpret_cast<CDirectorySAInterface*>(0xBC12C0);
    int   iReturnFileId;
    char* pReturnBuffer;
```

botder commented on Client/game_sa/CRenderWareSA.ClothesReplacing.cpp:
```diff
@@ -45,45 +48,65 @@ namespace
+
+            if (it != ms_OriginalStreamingSizesMap.end())
+            {
+                uint16 originalStreamingSize = it->second;
```
```suggestion
                std::uint16_t originalStreamingSize = it->second;
```

botder reviewed (changes requested):
Good job, it doesn't crash anymore

W3lac3 committed:
Merge remote-tracking branch 'upstream/master' into feature/replace-cj-clothe-model

W3lac3 committed:
Update due to code revision

- Change uint16 to std::uint16_t>
- Change the type of the variable g_clothesDirectory to auto

W3lac3 committed:
Merge remote-tracking branch 'upstream/master' into feature/replace-cj-clothe-model

W3lac3 committed:
Merge remote-tracking branch 'upstream/master' into feature/replace-cj-clothe-model

botder commented on Client/game_sa/CDirectorySA.cpp:
```diff
@@ -0,0 +1,50 @@
+    DirectoryInfoSA* entry = GetModelEntry(modelId);
+
+    if (!entry)
+        return false;
```
```suggestion
        return 0;
```

botder reviewed (changes requested):
[placeholder]

W3lac3 committed:
Merge remote-tracking branch 'upstream/master' into feature/replace-cj-clothe-model

W3lac3 committed:
Fix invalid return in CDirectorySAInterface::GetModelStreamingSize

Fernando-A-Rocha commented:
Congratulations @W3lac3 Do you need help documenting this on the wiki?

W3lac3 commented:
> Congratulations @W3lac3 Do you need help documenting this on the wiki?

If you could do the documentation, I'd appreciate it. I've been a bit busy.

Fernando-A-Rocha commented:
> > Congratulations @W3lac3 Do you need help documenting this on the wiki?
> 
> If you could do the documentation, I'd appreciate it. I've been a bit busy.

Which pages should be altered / created?

W3lac3 commented:
> Which pages should be altered / created?

The pages that need to be changed are: [clothes components](https://wiki.multitheftauto.com/wiki/Clothing_Component_IDs) and [engineLoadDFF](https://wiki.multitheftauto.com/wiki/EngineLoadDFF) (On this page there should be something showing that you can replace CJ's clothes, something like engineImportTXD wiki.)

Fernando-A-Rocha commented:
> > Which pages should be altered / created?
> 
> The pages that need to be changed are: [clothes components](https://wiki.multitheftauto.com/wiki/Clothing_Component_IDs) and [engineLoadDFF](https://wiki.multitheftauto.com/wiki/EngineLoadDFF) (On this page there should be something showing that you can replace CJ's clothes, something like engineImportTXD wiki.)

Done. Edited https://wiki.multitheftauto.com/wiki/Clothing_Component_IDs and added note to https://wiki.multitheftauto.com/wiki/EngineReplaceModel

FYI this was added in r22987

LiraGamer commented:
> > > Which pages should be altered / created?
> > 
> > 
> > The pages that need to be changed are: [clothes components](https://wiki.multitheftauto.com/wiki/Clothing_Component_IDs) and [engineLoadDFF](https://wiki.multitheftauto.com/wiki/EngineLoadDFF) (On this page there should be something showing that you can replace CJ's clothes, something like engineImportTXD wiki.)
> 
> Done. Edited https://wiki.multitheftauto.com/wiki/Clothing_Component_IDs and added note to https://wiki.multitheftauto.com/wiki/EngineReplaceModel
> 
> FYI this was added in r22987


you need change exemple has used 
**local dff = engineLoadTXD("custom_afrobeard.dff")**

correct is 
**local dff = engineLoadDFF("custom_afrobeard.dff")**
no ?

exemple used in : https://wiki.multitheftauto.com/wiki/Clothing_Component_IDs

Fernando-A-Rocha commented:
@LiraGamer Yes, you are right! I edited the page. thanks

LiraGamer commented:
![image](https://github.com/user-attachments/assets/f95daf4d-7ba6-4878-8e7e-2ffa94e96119)
im testing new fatures, im download cj female to teste.

to correct hair alpha change .txd to dxt5 or dxt3 the crash it's right
![image](https://github.com/user-attachments/assets/1301aa4a-6713-4428-876f-cdfd2881acb7)

its is correct ?

@TheNormalnij  @W3lac3 @botder

code used
`local dff = engineLoadDFF("afro.dff");
engineReplaceModel(dff, 30000);

local txd = engineLoadTXD("afro.txd")
engineImportTXD( txd, 30201 )`

W3lac3 commented:
> code used `local dff = engineLoadDFF("afro.dff"); engineReplaceModel(dff, 30000);
> 
> local txd = engineLoadTXD("afro.txd") engineImportTXD( txd, 30201 )`

I'm not sure why, but every time I edit txd using Magic TXD crashes, regardless of the compression I use. So try using TXD Workshop, because I used it while I was creating this system and everything worked normally. I also recommend asking questions about this function on the MTA discord.

PlatinMTA commented:
> code used `local dff = engineLoadDFF("afro.dff"); engineReplaceModel(dff, 30000);
> 
> local txd = engineLoadTXD("afro.txd") engineImportTXD( txd, 30201 )`

no crash while importing this, but if the texture is compressed (DXT1, 3, 5 or whatever) it will just look broken, so RGBA all the way... sadly that means bigger filesizes

![image](https://github.com/user-attachments/assets/102a8bb3-3cc3-40d2-a728-da2a2822c38d)
